### ISSUE TITLE
Linux/OSX PlatformNotSupportedException calling Socket.IOControl

### ISSUE THREAD

**Issue by luigiberrettini** [dotnet/corefx#28306]
PlatformNotSupportedException when [using IOControl](https://github.com/graffen/NLog.Targets.Syslog/blob/v5.0.0-beta-02/src/NLog.Targets.Syslog/MessageSend/Tcp.cs#L86) to set keepAlive
```csharp
tcp.Client.IOControl(IOControlCode.KeepAliveValues, keepAlive.ToByteArray(), null);
```
Please look at https://github.com/graffen/NLog.Targets.Syslog/issues/119 for more info

**Comment by stephentoub** [comment]
> PlatformNotSupportedException when using IOControl to set keepAlive

That's because it's only supported on Windows.  This IOControl code value is Windows-specific.

**Comment by karelz** [comment]
Closing per above comment

**Comment by luigiberrettini** [comment]
How can we pass KeepAlive on non Windows OSes?

Given a .NET Standard version, I thought that, building against a set of target frameworks that implement it, meant supporting all the features of that version on each platform the framework was installable on.

At the moment docs are quite confusing, because they do not warn develoers about platform specific issues and they suggest to see the Windows Sockets API even if the selected API is .NET Standard or .NET Core:
https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.iocontrol?view=netstandard-2.0#System_Net_Sockets_Socket_IOControl_System_Net_Sockets_IOControlCode_System_Byte___System_Byte___

Only looking at the IOControlCode enumeration documentation it is clear that it is a Windows specific thing, but it seems as the docs were taken from a Windows only world: nothing like “this control code is not supported on Linux or OSX; as for Windows it is only supported on Windows 2000 and later”

**Comment by stephentoub** [comment]
> How can we pass KeepAlive on non Windows OSes?

You should be able to use Socket.SetSocketOption with SocketOptionName.KeepAlive on all OSes.

> At the moment docs are quite confusing, because they do not warn develoers about platform specific issues and they suggest to see the Windows Sockets API even if the selected API is .NET Standard or .NET Core

cc: @terrajobst

**Comment by karelz** [comment]
@mairaw was incorporating our platform specific info from:
* https://github.com/dotnet/platform-compat
* https://github.com/dotnet/corefx/wiki/ApiCompat

However, I think this one is not called out. @luigiberrettini would you mind submitting a PR with docs changes to capture the fact? (@mairaw any recommnedations how?)

**Comment by luigiberrettini** [comment]
> You should be able to use Socket.SetSocketOption with SocketOptionName.KeepAlive on all OSes.

This does not allow to set [keepalive time and interval](https://github.com/graffen/NLog.Targets.Syslog/blob/v5.0.0-beta-02/src/NLog.Targets.Syslog/MessageSend/KeepAlive.cs) and I did not find any other way to do it

> @luigiberrettini would you mind submitting a PR with docs changes to capture the fact?

If you tell me the repo I can work on it

**Comment by stephentoub** [comment]
> This does not allow to set keepalive time and interval and I did not find any other way to do it

What API on Linux and macOS would?
cc: @tmds

**Comment by karelz** [comment]
Docs repo is: https://github.com/dotnet/docs (or just browse documentation and click the Edit button)
However, let's wait on guidance from @mairaw how to best capture platform differences in a unified way.

**Comment by wfurt** [comment]
Look at https://github.com/dotnet/corefx/issues/14237 @luigiberrettini 
We should drive that to completion. It is open for grabs.

**Comment by luigiberrettini** [comment]
> What API on Linux and macOS would?

Should be [this](http://tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/#setsockopt), but to avoid using platform specific PInvoke it should be wrapped by the framework

**Comment by luigiberrettini** [comment]
Superseded by #25040

**Comment by mairaw** [comment]
To answer your question @karelz, this is how we started doing this:
https://github.com/dotnet/docs/pull/3852 

In the future, we'll have conditional content, so the format might change, but this is how we started capturing the differences. And I'm open to feedback if anyone thinks there's a better way to do this.
